pennylane/gradients/parameter_shift_hessian.py

```diff
@@ -36,52 +35,118 @@
 )
 
 
+def _process_argnum(argnum, tape):
+    """Process the argnum keyword argument to ``param_shift_hessian`` from any of ``None``,
+    ``int``, ``Sequence[int]``, ``array_like[bool]`` to an ``array_like[bool]``."""
+    if argnum is None:
+        argnum = tape.trainable_params
+    elif isinstance(argnum, int):
+        argnum = [argnum]
+    if len(qml.math.shape(argnum)) == 1:
+        argnum = [i in argnum for i in range(tape.num_params)]
+        argnum = qml.math.tensordot(argnum, argnum, axes=0)
+    elif not (
+        len(qml.math.shape(argnum)) == 2
+        and qml.math.array(argnum).dtype == bool
+        and qml.math.allclose(qml.math.transpose(argnum), argnum)
+    ):
+        raise ValueError(f"Expected a symmetric Boolean array for argnum, but received {argnum}.")
+    return argnum
+
+
 def _collect_recipes(tape, argnum, diff_methods, diagonal_shifts, off_diagonal_shifts):
     r"""Extract second order recipes for the tape operations for the diagonal of the Hessian
     as well as the first-order derivative recipes for the off-diagonal entries.
     """
+    diag_argnum = list(qml.math.where(qml.math.diag(argnum))[0])
+    offdiag = argnum ^ qml.math.diag(qml.math.diag(argnum))
+    offdiag_argnum = list(qml.math.where(qml.math.any(offdiag, axis=0))[0])
```

Comment:
Ah, this is indeed a bit mean. Some hopefully useful remarks:
1. For the diagonal, we need second-order shift rules in a single parameter. Therefore we store in `diag_argnum` which of those shift rules we will need because `argnum[i, i]=True`.
2. For the off-diagonal entries, we will require first-order shift rules, that are then later combined into two-parameter shift rules. However, we only want to obtain and process each of these first-order rules only once, and we need the `i`-th parameter's rule only if _any_ of the entries in the `i`-th column (or row, the argnum mask is symmetric) is set to `True`. To achieve this, we first subtract the diagonal (Read `^` as `-`, which is not implemented for Boolean arrays), and then find the indices for which _any_ of the rows (`axis=0` in `qml.math.any`) contains a `True` value.

tl;dr: 
>For the off-diagonal argnum, it is still 1D but now contains the indices of the mask entries which where true in either of the 2 dimensions?

True, if you replace "in either of the 2 dimensions" by "in at least one row".

The conversion to index-based argnum could be avoided, making both `diag_argnum` and `offdiag_argnum` 1D Boolean arrays, if it is very annoying to read.